File: src/types/userSource.ts
/// 用户音源脚本信息
export interface UserSourceScript {
  id: string
  name: string
  version: string
  description: string
  author: string
  homepage: string
  script: string
  sources: Record<string, UserSourceInfo>
  allow_show_update_alert: boolean
  enabled: boolean
  created_at: number
  updated_at: number
}

/// 用户音源信息
export interface UserSourceInfo {
  type: 'music'
  actions: ('musicUrl' | 'lyric' | 'pic' | 'search')[]
  qualitys: ('128k' | '320k' | 'flac' | 'wav')[]
}

/// 音源类型
export type UserSourceType = 'music'

/// 音源操作类型
export type UserSourceAction = 'musicUrl' | 'lyric' | 'pic' | 'search'

/// 音质类型
export type Quality = '128k' | '320k' | 'flac' | 'wav'

/// 默认音质映射
export const DEFAULT_QUALITY_MAP: Record<Quality, Quality[]> = {
  '128k': ['128k'],
  '320k': ['128k', '320k'],
  'flac': ['128k', '320k', 'flac'],
  'wav': ['128k', '320k', 'flac', 'wav'],
}

File: src-tauri/src/types/user_source.rs
use serde::{Deserialize, Serialize};
use std::collections::HashMap;

/// 用户自定义音源脚本信息
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct UserSourceScript {
    pub id: String,
    pub name: String,
    pub version: String,
    pub description: String,
    pub author: String,
    pub homepage: String,
    pub script: String,
    pub sources: HashMap<String, UserSourceInfo>,
    pub allow_show_update_alert: bool,
    pub enabled: bool,
    #[serde(default = "default_user_source_priority")]
    pub priority: i32,
    pub created_at: i64,
    pub updated_at: i64,
}

fn default_user_source_priority() -> i32 {
    100
}

/// 用户音源信息
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct UserSourceInfo {
    #[serde(rename = "type")]
    pub source_type: String,
    pub actions: Vec<String>,
    pub qualitys: Vec<String>,
}

/// 用户音源状态
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct UserSourceState {
    pub status: bool,
    pub message: Option<String>,
    pub api_info: Option<UserSourceInfo>,
}

/// 脚本元数据（从 JS 文件注释中解析）
#[derive(Debug, Clone, Default)]
pub struct ScriptMetadata {
    pub name: String,
    pub version: Option<String>,
    pub description: Option<String>,
    pub author: Option<String>,
    pub homepage: Option<String>,
    pub sources: Option<HashMap<String, UserSourceInfo>>,
    pub allow_show_update_alert: Option<bool>,
}
